### PR TITLE
[Feature] Implement a simple TableReader for testing

### PR DESCRIPTION
**Pull request by banmoy**
## What type of PR is this：
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Realtime mv depends on `TableReader` for testing, so we first provide a simple implementation which is based on `TabletReader` and predicates. It only supports to read data from local tablets, and there is no guarantee for performance.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Review by decster — DISMISSED** [review]

**Review by murphyatwork — DISMISSED** [review]

File: be/src/storage/table_reader.cpp [hunk 1]
```diff
@@ -14,22 +14,121 @@
...
+#include "storage/tablet_manager.h"
+#include "storage/tablet_reader.h"
+#include "storage/union_iterator.h"
+
 namespace starrocks {
 
-TableReader::TableReader(const TableReaderParams& params) : _params(params) {}
+TableReader::TableReader(const TableReaderParams& params) : _params(params) {
+    for (auto& partition : _params.partition_param.partitions) {
+        for (auto& index : partition.indexes) {
```
Comment (ABingHuang) [inline be/src/storage/table_reader.cpp:30]: why collect all indexes' tablets? I think we should collect only base index's tablets?

File: be/src/storage/table_reader.cpp [hunk 2]
```diff
@@ -14,22 +14,121 @@
...
+        CHECK_EQ(_params.schema.field(i)->name(), _tablet_schema.field(i)->name());
+    }
+}
 
 TableReader::~TableReader() {}
 
 Status TableReader::multi_get(const Chunk& keys, const std::vector<std::string>& value_columns,
                               std::vector<bool>& found, Chunk& values) {
-    // TODO implement multi_get
+    DeferOp defer([&] { _obj_pool.clear(); });
```
Comment (LiShuMing) [inline be/src/storage/table_reader.cpp:51]: So this class is not thread-safe, only support one reader at one time?

File: be/src/storage/table_reader.cpp [hunk 3]
```diff
@@ -14,22 +14,121 @@
...
 
-TableReader::TableReader(const TableReaderParams& params) : _params(params) {}
+TableReader::TableReader(const TableReaderParams& params) : _params(params) {
+    for (auto& partition : _params.partition_param.partitions) {
+        for (auto& index : partition.indexes) {
+            for (int64_t tablet_id : index.tablets) {
+                TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true);
+                if (tablet != nullptr) {
+                    _local_tablets.push_back(tablet);
+                }
```
Comment (LiShuMing) [inline be/src/storage/table_reader.cpp:35]: What if tablet is nullptr? throw exception or add warning?

File: be/src/storage/table_reader.cpp [hunk 4]
```diff
@@ -14,22 +14,121 @@
...
+        _obj_pool.add(predicate);
+        predicates->push_back(predicate);
+    }
+}
+
+VectorizedSchema TableReader::_build_value_schema(const std::vector<std::string>& value_columns) {
+    VectorizedSchema schema;
+    for (auto& name : value_columns) {
+        VectorizedFieldPtr field = _tablet_schema.get_field_by_name(name);
+        CHECK(field != nullptr);
```
Comment (ABingHuang) [inline be/src/storage/table_reader.cpp:127]: avoid to use CHECK

File: be/src/storage/table_reader.cpp [hunk 5]
```diff
@@ -14,22 +14,121 @@
...
+            for (int64_t tablet_id : index.tablets) {
+                TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true);
+                if (tablet != nullptr) {
+                    _local_tablets.push_back(tablet);
+                }
+            }
+        }
+    }
+
+    _tablet_schema = ChunkHelper::convert_schema_to_format_v2(_local_tablets[0]->tablet_schema());
```
Comment (LiShuMing) [inline be/src/storage/table_reader.cpp:40]: what if _local_tablets'size is empty?

File: be/src/storage/table_reader.cpp [hunk 6]
```diff
@@ -14,22 +14,121 @@
...
 
 Status TableReader::multi_get(const Chunk& keys, const std::vector<std::string>& value_columns,
                               std::vector<bool>& found, Chunk& values) {
-    // TODO implement multi_get
+    DeferOp defer([&] { _obj_pool.clear(); });
+    VectorizedSchema value_schema = _build_value_schema(value_columns);
+    ChunkPtr read_chunk = ChunkHelper::new_chunk(value_schema, 1);
+    std::vector<const ColumnPredicate*> predicates;
+    Status status;
+    found.clear();
```
Comment (LiShuMing) [inline be/src/storage/table_reader.cpp:56]: resize `found`  to keys'size at first? or reserve its size?

File: be/src/storage/table_reader.cpp [hunk 7]
```diff
@@ -14,22 +14,121 @@
...
+        if (!status.ok()) {
+            return status;
+        }
+        tablet_readers.push_back(reader);
+    }
+
+    return new_projection_iterator(value_schema, new_union_iterator(std::move(tablet_readers)));
+}
+
+void TableReader::_build_get_predicates(DatumTuple& tuple, std::vector<const ColumnPredicate*>* predicates) {
```
Comment (LiShuMing) [inline be/src/storage/table_reader.cpp:112]: `predicates` first reserve?

File: be/src/storage/table_reader.cpp (hunk 4 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: return error status instead of CHECK

File: be/src/storage/table_reader.cpp (hunk 1 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: the user will ensure there will be one index in each partition_param, and just read the first index

File: be/src/storage/table_reader.cpp (hunk 2 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: not share ObjectPool across multiple requests to support concurrency

File: be/src/storage/table_reader.cpp (hunk 5 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: add protection

File: be/src/storage/table_reader.cpp (hunk 7 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: Done

File: be/src/storage/table_reader.cpp (hunk 6 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: Done

File: be/src/storage/table_reader.cpp (hunk 3 above)
Comment (banmoy) [reply be/src/storage/table_reader.cpp]: If there is no local tablets, multi_get will return not found, and scan will return empty iterator

**Comment by banmoy** [comment]
run starrocks_admit_test

**Review by decster — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]
LGTM

**Comment by banmoy** [comment]
run starrocks_admit_test

**Comment by banmoy** [comment]
run starrocks_admit_test

**Pull request merged**